### ISSUE TITLE
support pyright?

### ISSUE THREAD

**Issue by wrvsrx** [ramonhagenaars/nptyping#61]
After installing v2.0.0, I found it works with mypy while it doesn't work with pyright.

For example, following code pass examimation by mypy, but it can't pass that by pyright.

```python
from nptyping import Shape
```

Mypy gives no error, pyright gives that

![image](https://user-images.githubusercontent.com/42770726/162426627-58a01012-37c5-4001-b90f-be41384116f0.png)

**Comment by ramonhagenaars** [comment]
Hi @wrvsrx ,

You are correct that it does not work with pyright at the moment. For the `2.0.0` release, the primary focus has been layed on MyPy. 

I'm not familiar with Pyright, so I don't know the implications right now, but I think it's interesting to support more than just one static type checker.

I am also interested to know how popular this feature would be - @readers: raise a 👍 if you are using Pyright.

**Comment by leycec** [comment]
**Drive-by commentary:** sadly for us all, Pyright is kinda popular. Why? Because the [most popular Python plugin for the most popular IDE (...*yup, it's Microsoft Pylance **+** VSCode*)](https://marketplace.visualstudio.com/items?itemName=ms-python.vscode-pylance) is driven under-the-hood by Pyright.

Personally, I'm Vim **+** mypy for the win; if it isn't painfully low-level and fully community-owned, I'm out. Pragmatically, market realities suggest that everyone disagrees with my personal life choices.

Really, the only relevant question is: how do real-world data scientists type-check? If my wife is a representative sample, the answer is **Jupyter + [nbQA](https://github.com/nbQA-dev/nbQA).** nbQA extends mypy (but *not* Pyright) support to Jupyter.

## tl;dr

Although everyone *usually* cares about Pyright (because Pylance), nobody who NumPy's cares about Pyright (because Jupyter). Data scientists are a different breed of Pythonista.

## Actually...

**This issue can be safely closed.** Hear me out, unruly mob!

[Pyright has *no* plugin API and no intention to ever define one](https://github.com/microsoft/pyright/issues/607). Pyright devs are [purist PEP zealots who mistakenly believe that plugins for static type-checkers encourage non-standard typing semantics](https://github.com/microsoft/pyright/issues/607#issuecomment-873467941):

> **We will not be providing a proprietary plug-in model akin to the one provided by mypy.** There are too many problems with such a model. It doesn't work across type checkers, maintenance of plugins is a problem, distribution and installation of plugins is a problem (since users need to know they exist and how to configure them), and there are potential security issues with plugins that are downloaded and run on users' machines.

> In general, we want to encourage authors of new Python libraries (or new versions of existing Python libraries) to take into account static type checking when designing their interface contracts. **Non-standard semantics and dynamic behaviors are antithetical to static type checking** and work against efforts to improve the Python coding experience with tools such as language servers and type checkers.

They're wrong, because fundamentalists usually are. Right? When fundamentalism collides with the physical reality, physical reality wins. But you're not going to convince them otherwise, because fundamentalists entrench when you confront them with factual counter-evidence.

Myself and other @beartype collaborators have tried opening discussions with Pyright devs on a variety of topics. It doesn't work, because Pyright devs are disinterested in having those discussions. It is what it is.

In short, nothing can be done. There's nowhere for this issue to go, so this issue should probably be quietly laid to rest. :headstone:

**Comment by ramonhagenaars** [comment]
Thank you for this elaborate drive-by, @leycec.

Given your experience, I have little confidence in the feasability of this feature. It was one thing to make `nptyping` "MyPyable", but multiple static type checkers with their own quirks and opinions... and without a plug-in system to cheat our way out in case of emergency...

I think it's a bitter but fair conclusion that this feature is likely not gonna happen. Closing it.
